### ISSUE TITLE
Improve exception handling for swift

### ISSUE THREAD

**Issue by thevery** [JetBrains/kotlin-native#1608]
```
    @konan.Throws(Exception::class)
    fun hello1() {
        throw Exception()
    }

    @konan.Throws(Exception::class)
    fun hello2(): String {
        throw Exception()
    }

    @konan.Throws(Exception::class)
    fun hello3(): String? {
        throw Exception()
    }

    @konan.Throws(Exception::class)
    fun hello4(s: String): Boolean {
        throw Exception()
    }
```
generates following obj-c
```
-(BOOL)hello1AndReturnError:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello1());
-(NSString * _Nullable)hello2AndReturnError:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello2());
-(BOOL)hello3AndReturnResult:(NSString * _Nullable * _Nullable)result error:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello3(result:));
-(BOOL)hello4S:(NSString *)s result:(BOOL * _Nullable)result error:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello4(s:result:));
```
and swift
```
    open func hello1() throws
    open func hello2() throws -> String
    open func hello3(result: AutoreleasingUnsafeMutablePointer<NSString?>?) throws
    open func hello4(s: String, result: UnsafeMutablePointer<ObjCBool>?) throws

```
`hello1` and `hello2` works as intended, but `hello3` and `hello3` should be the following:
```
-(NSString * _Nullable)hello3WithError:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello3);
-(BOOL)hello4WithS:(NSString *)s error:(NSError * _Nullable * _Nullable)error NS_SWIFT_NAME(hello4(S:));
```
```
    open func hello3() throws -> String?
    open func hello4(s: String) throws -> Boolean
```

**Comment by SvyatoslavScherbina** [comment]
There is a reason behind the behaviour observable with `hello3` and `hello4`.
According to Cocoa conventions, if error-producing method returns `NO` or `nil`, then an error have occurred. So caller can't distinguish between the case when a Kotlin exception has been thrown and the case when Kotlin method returns `false` or `null`.
And we don't provide separate API for Swift yet.

Also current behaviour shouldn't bring noticeable inconveniences, because in both Swift and Objective-C it is possible to pass reference to local variable for `result` parameter. This fact was also considered when we were designing `@Throws`.

**Comment by SvyatoslavScherbina** [comment]
Fixed in 1.4-M1.
